src/daemon/daemon.cpp

```diff
@@ -386,10 +395,91 @@ auto connect_rpc(mp::DaemonRpc& rpc, mp::Daemon& daemon)
     QObject::connect(&rpc, &mp::DaemonRpc::on_start, &daemon, &mp::Daemon::start, Qt::BlockingQueuedConnection);
     QObject::connect(&rpc, &mp::DaemonRpc::on_stop, &daemon, &mp::Daemon::stop, Qt::BlockingQueuedConnection);
     QObject::connect(&rpc, &mp::DaemonRpc::on_suspend, &daemon, &mp::Daemon::suspend, Qt::BlockingQueuedConnection);
+    QObject::connect(&rpc, &mp::DaemonRpc::on_restart, &daemon, &mp::Daemon::restart, Qt::BlockingQueuedConnection);
     QObject::connect(&rpc, &mp::DaemonRpc::on_delete, &daemon, &mp::Daemon::delet, Qt::BlockingQueuedConnection);
     QObject::connect(&rpc, &mp::DaemonRpc::on_umount, &daemon, &mp::Daemon::umount, Qt::BlockingQueuedConnection);
     QObject::connect(&rpc, &mp::DaemonRpc::on_version, &daemon, &mp::Daemon::version, Qt::BlockingQueuedConnection);
 }
+
+template<typename Targets, typename InstanceMap>
+grpc::Status validate_targets(const Targets& targets,
+                              const InstanceMap& vms,
+                              const InstanceMap& deleted)
+{
+    fmt::memory_buffer errors;
+    for(const auto& name : targets)
+        if(vms.find(name) == std::cend(vms))
+        {
+            if(deleted.find(name) == std::cend(deleted))
+                fmt::format_to(errors, "instance \"{}\" does not exist\n", name);
+            else
+                fmt::format_to(errors, "instance \"{}\" is deleted\n", name);
+        }
+
+    return errors.size() ? grpc_status_for(errors) : grpc::Status::OK;
+}
+
+template<typename Targets, typename InstanceMap>
+auto process_targets(const Targets& targets,
+                     const InstanceMap& vms,
+                     const InstanceMap& deleted)
+-> std::pair<std::vector<typename Targets::value_type>, grpc::Status>
+{   // TODO: use this in commands that currently duplicate the same kind of code
+    auto st = validate_targets(targets, vms, deleted);
+    auto tgts = std::vector<typename Targets::value_type>{};
```

Comment:
Also, maybe use `valid_instances` instead of `tgts` here.

Reply:
Done.